### ISSUE TITLE
Unions inside return types should generate all variants of that type

### ISSUE THREAD

**Issue by Thibaulltt** [astral-sh/ty#3012]
### Summary

Link to a playground: https://play.ty.dev/1820887f-a950-48f6-813c-5cf02c713cff

If I have two functions, one of which might return a list of `Foo` and one of which might return a list of `Bar`, and I call them from a `foo_or_bar(...)` function which returns one of those items here is how I expect to type the return type hint of `foo_or_bar`:

```py
def foo(...) -> List[Foo]:
    # ...
    return ...

def bar(...) -> List[Bar]:
    # ...
    return ...

def foo_or_bar(...) -> List[Union[Foo, Bar]]:
    # ...
    return ...
```

But `ty` seems to think returning a `List[Foo]` items do not match the return type of the last function. Shouldn't `ty` expand all "variants" of the return type and see if the inferred type of the return value match one of them ?

If I return a `List[Optional[Foo]]` from a function, `ty` does not complain even though I remember reading somewhere `Optional` is a simple case of `Union[_T, None]`...

### Version

ty 0.0.18

**Comment by sharkdp** [comment]
> But `ty` seems to think returning a `List[Foo]` items do not match the return type of the last function

And it shouldn't. `List[Foo]` is not assignable to `List[Union[Foo, Bar]]` because `List` is [invariant](https://typing.python.org/en/latest/spec/generics.html#variance) in its type parameter (see the first paragraph in the linked section of the typing docs). You probably want `Union[List[Foo], List[Bar]]`, i.e. either a list of `Foo`s, or a list of `Bar`s (not a list that can contain `Foo`s and `Bar`s).

> If I return a `List[Optional[Foo]]` from a function, `ty` does not complain even though I remember reading somewhere `Optional` is a simple case of `Union[_T, None]`...

Yes, `Optional[Foo]` is the same thing as `Foo | None` or `Union[Foo, None]`. Can you show us the full example which shows this inconsistency?

**Comment by Thibaulltt** [comment]
Here's a redacted example (research code, not ready for showing yet):

```py
def _internal_loss_foo(stdout: str) -> List[List[Tuple[int, float]]]:
    # Read data from stdout and format it.
    grouped_losses = ...
    # ...
	return grouped_losses


def _internal_loss_bar(out: str) -> List[Tuple[int, float]]:
	# Read data from stdout and format it.
	grouped_losses = ...
	# ...
	return grouped_losses


def try_extract_internal_losses(result: Result) -> List[Union[List[Tuple[int, float]], Tuple[int, float]]]:
	"""
	Attempt to extract losses either using Foo or Bar.

	:param Result result: Contains the standard output to analyze.
	:returns: Either a list of ``[(inner_iteration_0, loss), ..., (inner_iteration_n, loss)]`` or a list of ``(iteration, loss)``.
	"""

	if result.parameters.solver in [...]:  # if we are using a "foo" solver, many implementations but all parsed here.
		foo_losses = _internal_loss_foo(result.stdout)
		if len(foo_losses) > 0:
			_info(f'Found {[len(losses) for losses in foo_losses]} iterations of Foo (-like) losses inside.')
			return foo_losses

	if result.parameters.solver == 'bar':
		bar_losses = _internal_loss_bar(result.stdout)
		if len(bar_losses) > 0:
			_info(f'Found {len(bar_losses)} iterations of Bar losses inside.')
			return sinkhorn_losses

	_error(f'Unrecognized solver: {result.parameters.solver.name}')
	return []
```

And here's an example where the `Optional` version is correctly inferred:

```py
T = TypeVar('T')

def convert_tupled_items_to_list(losses: List[Tuple[int, T]]) -> List[Optional[T]]:
	""" This function "flattens" a list of tuples containing ``(iteration, <something>)`` to a possibly non-contiguous list of ``<something>s``.
	:param losses: The list of tuples to convert, each with its desired list index in the 0-th index position.
	:returns: A list of possibly ``None`` ``<something>s``. """
	_max = losses[-1][0]
	converted = [None] * _max

	for iteration, loss in losses:
		converted[iteration] = loss

	return converted

# later...
losses = try_extract_internal_losses(...)  # This call returns the List[List[Tuple[int, float]]] variant
for ..., loss in zip(..., losses)
    foo_store.losses = convert_tupled_items_to_list(loss)  # So, here loss = List[Tuple[int, float]]
    #         ^ losses is defined as a List[Optional[float]] in the foo_store type.
```

> You probably want `Union[List[Foo], List[Bar]]`

That's an alternative, but I didn't want to expand all variants manually :(

**Comment by sharkdp** [comment]
> > You probably want `Union[List[Foo], List[Bar]]`
> 
> That's an alternative, but I didn't want to expand all variants manually :(

It's not an alternative, it's a different type — the correct type for annotating a function that can return either a list of `Foo`'s, or a list of `Bar`'s.

`List[Union[Foo, Bar]]` is something else. It's a type that represents lists that can contain `Foo`'s, or `Bar`'s, *or both*.

And since `List` is invariant in its argument, `List[Foo]` is not assignable to `List[Union[Foo, Bar]]`. Otherwise, you could pass a `List[Foo]` to a function that accepts `List[Union[Foo, Bar]]`, and internally, that function could type-safely modify your list and append a `Bar`, and you as the caller would think that it's still a list of `Foo`'s.

> And here's an example where the Optional version is correctly inferred:

I think that example only demonstrates that you can assign `List[Optional[T]]` (the return type of the `convert_tupled_items_to_list(loss)` call) to `List[Optional[T]]` (the declared type of `foo_store.losses`), but I don't see how that is relevant to the other discussion?

**Comment by Thibaulltt** [comment]
> I think that example only demonstrates that you can assign `List[Optional[T]]` (the return type of the `convert_tupled_items_to_list(loss)` call) to `List[Optional[T]]` (the declared type of `foo_store.losses`), but I don't see how that is relevant to the other discussion?

It was an example where I thought `ty` correctly inferred that the losses variable above _did_ contain either `List[Tuple[int, float]]` or `Tuple[int, float]` but upon closer examination, I see the type is `Unknown` (probably because of the expansion when iterating on `zip(...)`).

I think my fundamental mistake before opening this issue was misunderstanding variance in the `List` type (or its lack thereof, to be precise). It's not pretty, but I'll have to manually expand all type variants I can return in those functions... Thanks anyway!
